### PR TITLE
[incubator/elasticsearch] Using DNS discovery instead of kubernetes plugin

### PR DESCRIPTION
**Pull request by rendhalver**
Recreating #2889. see that PR for original intent and discussion

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by rendhalver** [comment]
I think this is actually working but it's taking longer than the tests are checking for.
I spun one up locally and it eventually finished.
I will do some checking to see how long it takes and if it's possible to speed it up and post my findings.

File: incubator/elasticsearch/templates/client-deployment.yaml [hunk 1]
```diff
@@ -21,7 +21,8 @@ spec:
 {{ toYaml .Values.client.podAnnotations | indent 8 }}
         {{- end }}
     spec:
-      serviceAccountName: {{ template "elasticsearch.fullname" . }}
+      securityContext:
+        fsGroup: 1000
```
Comment (timstoop) [inline incubator/elasticsearch/templates/client-deployment.yaml:25]: Won't this break old clusters as well? I'm not entirely sure how Kubernetes does this, if it does a chown on the Volumes or something. Have you tested the described upgrade, by any chance?

File: incubator/elasticsearch/README.md [hunk 2]
```diff
@@ -1,7 +1,13 @@
 # Elasticsearch Helm Chart
 
-This chart is based on the [centerforopenscience/elasticsearch](https://hub.docker.com/r/centerforopenscience/elasticsearch/) image which comes with Fabric8's great [kubernetes discovery plugin](https://github.com/fabric8io/elasticsearch-cloud-kubernetes) for Elasticsearch.
- 
+This chart uses a standard Docker image of Elasticsearch (hub.docker.com/r/blacktop/elasticsearch) and uses a service pointing to the master's transport port for service discovery.
+Elasticsearch does not communicate with the Kubernetes API, hence no need for RBAC permissions.
+
+## Warning for previous users
```
Comment (timstoop) [inline incubator/elasticsearch/README.md:6]: Are we certain this is tested thoroughly? If not, we might want to consider recommending people to create a new cluster instead. I had a (test) cluster fail during while I tested a very early version of these new changes, wouldn't want that to happen to anyone depending on this chart.

**Comment by timstoop** [comment]
Aside from the few comments I made, I'd love to see this merged.

File: incubator/elasticsearch/templates/client-deployment.yaml (hunk 1 above)
Comment (rendhalver) [reply incubator/elasticsearch/templates/client-deployment.yaml:25]: I haven’t tested an upgrade myself.
I will see how that goes.

And now I am wondering how we could do that with CI. :)

**Comment by rendhalver** [comment]
Me too.
I will see about getting an upgrade tested when I get a chance.

File: incubator/elasticsearch/templates/client-deployment.yaml (hunk 1 above)
Comment (timstoop) [reply incubator/elasticsearch/templates/client-deployment.yaml:25]: That would be a really cool feature...

**Comment by rendhalver** [comment]
I just tried upgrading a cluster built with the current release and it didn't get far.
This is what the pod list looks like before I ran the upgrade.
```
07:50 $ kubectl --context=docker-for-desktop --namespace=es-test get pods
NAME                                               READY     STATUS    RESTARTS   AGE
crazy-horse-elasticsearch-client-76b7c4d4c-8hjmx   1/1       Running   0          8m
crazy-horse-elasticsearch-client-76b7c4d4c-rqxd8   1/1       Running   0          8m
crazy-horse-elasticsearch-data-0                   1/1       Running   0          8m
crazy-horse-elasticsearch-data-1                   1/1       Running   0          7m
crazy-horse-elasticsearch-master-0                 1/1       Running   0          8m
crazy-horse-elasticsearch-master-1                 1/1       Running   0          6m
crazy-horse-elasticsearch-master-2                 1/1       Running   0          6m
```
and this a few minutes after
```
07:58 $ kubectl --context=docker-for-desktop --namespace=es-test get pods
NAME                                                READY     STATUS     RESTARTS   AGE
crazy-horse-elasticsearch-client-766587bbbf-z9scq   0/1       Init:0/1   0          7m
crazy-horse-elasticsearch-client-76b7c4d4c-8hjmx    1/1       Running    0          18m
crazy-horse-elasticsearch-client-76b7c4d4c-rqxd8    1/1       Running    0          18m
crazy-horse-elasticsearch-data-0                    1/1       Running    0          18m
crazy-horse-elasticsearch-data-1                    1/1       Running    0          16m
crazy-horse-elasticsearch-master-0                  1/1       Running    0          18m
crazy-horse-elasticsearch-master-1                  1/1       Running    0          16m
crazy-horse-elasticsearch-master-2                  1/1       Running    0          15m
```

It looks like it's attempting to do something but hasn't completed.

I am wondering if the removal of the RBAC resources has something to do with it.

**Comment by rendhalver** [comment]
So upgrading to this may not work.
So I guess the big question is how do we maintain data integrity if we are recommending people rebuild their clusters?

**Comment by timstoop** [comment]
Any chance you can check the logs of that init container to see what it's waiting for?

**Comment by monotek** [comment]
As long as the image version is configurable i don't think there is an easy auto update option via init container. I'd would just give an update note in the readme.

**Comment by rendhalver** [comment]
```
07:59 $ kubectl --context=docker-for-desktop --namespace=es-test logs -f crazy-horse-elasticsearch-client-766587bbbf-z9scq
Error from server (BadRequest): container "elasticsearch" in pod "crazy-horse-elasticsearch-client-766587bbbf-z9scq" is waiting to start: PodInitializing
```

**Comment by timstoop** [comment]
`kubectl --context=docker-for-desktop --namespace=es-test logs -f crazy-horse-elasticsearch-client-766587bbbf-z9scq -c sysctl` should work, I think, that's the first init container configured. Might actually be the `chown` container as well, if there was a lot of data in there already.

**Comment by rendhalver** [comment]
This will probably help
```
14:06 $ kubectl --context=docker-for-desktop --namespace=es-test get pod crazy-horse-elasticsearch-client-766587bbbf-z9scq -o yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: 2018-05-23T11:52:37Z
  generateName: crazy-horse-elasticsearch-client-766587bbbf-
  labels:
    app: elasticsearch
    component: client
    pod-template-hash: "3221436669"
    release: crazy-horse
  name: crazy-horse-elasticsearch-client-766587bbbf-z9scq
  namespace: es-test
  ownerReferences:
  - apiVersion: extensions/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: crazy-horse-elasticsearch-client-766587bbbf
    uid: c9ee358e-5e7f-11e8-b2f5-025000000001
  resourceVersion: "3071093"
  selfLink: /api/v1/namespaces/es-test/pods/crazy-horse-elasticsearch-client-766587bbbf-z9scq
  uid: c9f3b26d-5e7f-11e8-b2f5-025000000001
spec:
  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchLabels:
              app: elasticsearch
              component: client
              release: crazy-horse
          topologyKey: kubernetes.io/hostname
        weight: 1
  containers:
  - env:
    - name: NODE_DATA
      value: "false"
    - name: NODE_INGEST
      value: "false"
    - name: NODE_MASTER
      value: "false"
    - name: DISCOVERY_SERVICE
      value: crazy-horse-elasticsearch-discovery.es-test.svc.cluster.local
    - name: PROCESSORS
      valueFrom:
        resourceFieldRef:
          divisor: "0"
          resource: limits.cpu
    - name: ES_JAVA_OPTS
      value: -Djava.net.preferIPv4Stack=true -Xms512m -Xmx512m
    - name: MINIMUM_MASTER_NODES
      value: "2"
    image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.1
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /_cluster/health?wait_for_status=yellow
        port: 9200
        scheme: HTTP
      initialDelaySeconds: 90
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    name: elasticsearch
    ports:
    - containerPort: 9200
      name: http
      protocol: TCP
    - containerPort: 9300
      name: transport
      protocol: TCP
    readinessProbe:
      failureThreshold: 3
      httpGet:
        path: /_cluster/health?wait_for_status=yellow
        port: 9200
        scheme: HTTP
      initialDelaySeconds: 5
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "1"
      requests:
        cpu: 25m
        memory: 512Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /usr/share/elasticsearch/config/elasticsearch.yml
      name: config
      subPath: elasticsearch.yml
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: crazy-horse-elasticsearch-token-swxw7
      readOnly: true
  dnsPolicy: ClusterFirst
  initContainers:
  - command:
    - sysctl
    - -w
    - vm.max_map_count=262144
    image: busybox
    imagePullPolicy: Always
    name: sysctl
    resources: {}
    securityContext:
      privileged: true
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: crazy-horse-elasticsearch-token-swxw7
      readOnly: true
  nodeName: docker-for-desktop
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 1000
  serviceAccount: crazy-horse-elasticsearch
  serviceAccountName: crazy-horse-elasticsearch
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - configMap:
      defaultMode: 420
      name: crazy-horse-elasticsearch
    name: config
  - name: crazy-horse-elasticsearch-token-swxw7
    secret:
      defaultMode: 420
      secretName: crazy-horse-elasticsearch-token-swxw7
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-05-23T11:52:37Z
    message: 'containers with incomplete status: [sysctl]'
    reason: ContainersNotInitialized
    status: "False"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-05-23T11:52:37Z
    message: 'containers with unready status: [elasticsearch]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-05-23T11:52:37Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: docker.elastic.co/elasticsearch/elasticsearch-oss:6.1.1
    imageID: ""
    lastState: {}
    name: elasticsearch
    ready: false
    restartCount: 0
    state:
      waiting:
        reason: PodInitializing
  hostIP: 192.168.65.3
  initContainerStatuses:
  - image: busybox
    imageID: ""
    lastState: {}
    name: sysctl
    ready: false
    restartCount: 0
    state:
      waiting:
        reason: PodInitializing
  phase: Pending
  qosClass: Burstable
  startTime: 2018-05-23T11:52:37Z
```

**Comment by rendhalver** [comment]
It looks like most of the containers are still initialising. :(

File: incubator/elasticsearch/Chart.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 name: elasticsearch
 home: https://www.elastic.co/products/elasticsearch
-version: 0.4.9
+version: 0.6.1
```
Comment (unguiculus) [inline incubator/elasticsearch/Chart.yaml]: 0.5.0?

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/elasticsearch/Chart.yaml (hunk 3 above)
Comment (rendhalver) [reply incubator/elasticsearch/Chart.yaml]: We had a few version increments on this locally and this was the next biggest version.

**Comment by rendhalver** [comment]
I noticed the appVersion is still 5.4.
I think we should increase that to 6.1.1.

**Comment by abhishenator** [comment]
Hi @rendhalver, Thanks for this PR. Much needed. I picked up changes from this PR and deployed. I m seeing discovery error:

```
java.net.UnknownHostException: es-test-elasticsearch-discovery.abhi-test-pipeline.svc.cluster.local
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280) ~[?:1.8.0_151]
	at java.net.InetAddress.getAllByName(InetAddress.java:1192) ~[?:1.8.0_151]
	at java.net.InetAddress.getAllByName(InetAddress.java:1126) ~[?:1.8.0_151]
	at org.elasticsearch.transport.TcpTransport.parse(TcpTransport.java:917) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.transport.TcpTransport.addressesFromString(TcpTransport.java:872) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.transport.TransportService.addressesFromString(TransportService.java:699) ~[elasticsearch-6.1.1.jar:6.1.1]
	at org.elasticsearch.discovery.zen.UnicastZenPing.lambda$null$0(UnicastZenPing.java:213) ~[elasticsearch-6.1.1.jar:6.1.1]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_151]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:568) [elasticsearch-6.1.1.jar:6.1.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_151]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
[2018-05-25T17:04:13,043][WARN ][o.e.d.z.ZenDiscovery     ] [es-test-elasticsearch-master-0] not enough master nodes discovered during pinging (found [[Candidate{node={es-test-elasticsearch-master-0}{Bld3gg-eTCufU4ZxhUrYoA}{y2Jbvc2QSJ2uZzR_bNbhqQ}{10.1.1.157}{10.1.1.157:9300}, clusterStateVersion=-1}]], but needed [2]), pinging again
[2018-05-25T17:04:13,044][WARN ][o.e.d.z.UnicastZenPing   ] [es-test-elasticsearch-master-0] failed to resolve host [es-test-elasticsearch-discovery.abhi-test-pipeline.svc.cluster.local]
```

Please let me know if you have any suggestion. Thanks.

**Comment by abhishenator** [comment]
Follow up on my previous comment. I realized that my cluster domain is not `cluster.local` but `cluster.<my k8s cluster name>`. That leads to problem in discovery if I use default chart values. We usually expect the default values to work on chart deployment which is not the case here.

I would suggest to change `{{ template "elasticsearch.fullname" . }}-discovery.{{ .Release.Namespace }}.svc.{{ .Values.cluster.kubernetesDomain }}` to `{{ template "elasticsearch.fullname" . }}-discovery` in templates/client-deployment.yaml, templates/data-statefulset.yaml, templates/master-statefulset.yaml, as the default lookup happens on current namespace and cluster domain where the chart is installed anyway. So, adding that information seems redundant and actually causes issues using default values which might not apply on user's environment.

**Comment by Flowkap** [comment]
that my cluster domain is not cluster.local but cluster.<my k8s cluster name>

I recently connected on that on another ticket. Many people seem to not know that cluster.local cannot be expected and neither is needed. The default DNS names don't use the cluster name usually.

**Comment by rendhalver** [comment]
Good idea @abhishenator I will switch it to namespace specific DNS.

**Comment by rendhalver** [comment]
Well that didn't work.
I am going to switch it back.
If you have a different base domain you can adjust `Values.cluster.kubernetesDomain`.

**Comment by rendhalver** [comment]
I am also going to try building a new cluster with 6.1.1 using the current Chart and then upgrade it with this.
Just to see if this actually works with the same version.
Upgrading ES to 6.1.1 and switching the cluster discovery at the same time obviously doesn't work.

File: incubator/elasticsearch/README.md [hunk 4]
```diff
@@ -66,7 +66,6 @@ The following table lists the configurable parameters of the elasticsearch chart
 | `image.tag`                          | Container image tag                                                 | `5.4`                                |
```
Comment (unguiculus) [inline incubator/elasticsearch/README.md:67]: 6.1.1

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/elasticsearch/README.md (hunk 4 above)
Comment (rendhalver) [reply incubator/elasticsearch/README.md:67]: Thanks. I will get that fixed.

**Comment by abhishenator** [comment]
The discovery works with cluster domain removed from the the service name with elasticsearch-oss/6.1.1. Surprisingly, it does not work with the other image: elasticsearch/5.3.3. This looks a bit puzzling. This could also mean that the new discovery patch does not work with other images like elasticsearch/5.3.3. (I need to test this.)

**Comment by abhishenator** [comment]
Looks like I m encountering a bootstrapping problem with discovery service. I m not sure how the current logic would work. Clients, datanodes and masters node are relying on discovery service to discover the cluster. The discovery service(hosted by master nodes) is relying on master nodes readiness for the service endpoints to be UP. Looks like this is a deadlock situation. The master's readiness probe looks at the local state(than the cluster state). This hopefully should work but it doesnt seem like we have it working with other images.

**Comment by Vince-Cercury** [comment]
I've tested creating a cluster on kube-aws and Kubernetes 1.9.3 and I didn't face any issue. Good job!
I'm glad we don't have to setup the ClusterRole anymore.

I'm happy to use `elasticsearch-oss/6.1.1`. I've tested with JanusGraph and Cassandra, they are able to use it (any 5.x or 6.x works).

Thanks guys!

**Comment by abhishenator** [comment]
Hi guys, Can we also test with elastic search images other than `elasticsearch-oss/6.1.1`? Example: `elasticsearch:5.*`, etc. I faced problems with other images.

**Comment by timstoop** [comment]
Although I'm not against testing those older versions per se, why would we want to? What would be a reason to use an old version of the app but a newer version of the chart? I think it may become a bit of a hassle if we need to start supporting older versions of all applications as well. I much rather choose a version to support for a particular Chart version (that's what the `appVersion` key is for, right?).

**Comment by abhishenator** [comment]
This PR mostly aims to remove discovery dependency on the k8s plugin. I was assuming that was the only scope of this PR without causing regressions in support of earlier versions. I m fine if we  chose to break that. Though, there may be enough people who are still using ES 5.* and don't want to rely on k8s plugin for discovery as they faced issues with that. Those people have no choice but to look out for other charts. This being official helm repo for popular apps should aim for minimal regressions wherever possible. In any case, we should also test with `elasticsearch:6.*` which also didn't work for me atleast. Would be great if we can confirm this works with few other 6.* images if we only plan to support 6.* from now on.

File: incubator/elasticsearch/templates/client-deployment.yaml (hunk 1 above)
Comment (rendhalver) [reply incubator/elasticsearch/templates/client-deployment.yaml:25]: I should go look at the CI setup sometime.

**Comment by rendhalver** [comment]
I just built a new test cluster with the Chart from current master and then upgraded that with this PR.
It had the same issue I was having before where it started one client pod which didn't finish initialising.
I kept the same image for the new nodes and kept appversion at 5.4 just to rule that out.

The events for the namespace seem to be saying the current pods are having trouble getting to a secret mount which looks like it might be the old token that got removed when we removed the RBAC resources.

My guess is the we can't join the new node to the cluster because of how the current cluster does it's discovery.

So we may have to recommend people rebuild their clusters unless we can work out a way to get around that.
Any ideas?

**Comment by timstoop** [comment]
I'm in favour of doing a backward incompatible change to get this change through. Note it clearly in the docs and increase the major version. The previous charts had enough issues to warrant a complete rebuild of the cluster. If we can make it so that the old data volumes can still be used, it actually doesn't really have to be a very big issue, just some downtime. You can actually set up the new cluster next to the old one and migrate data over. This requires a one line setting in the ES config. We have a script available to help with such migrations, I can check if we can provide that (I'll probably need to tweak it to make it a bit more generic/customizable), if people are interested?

**Comment by rendhalver** [comment]
Sounds good @timstoop.
I will adjust the README and increase the major version.

**Comment by abhishenator** [comment]
Thanks guys for this PR. This is really helpful. Btw, I was able to get 5.* working. The main problem was that Basic Auth was needed in the readiness probes. We can accomplish this using https://stackoverflow.com/questions/33484942/how-to-use-basic-authentication-in-a-http-liveness-probe-in-kubernetes/33488076.

We don't need to support this. Posting this in the hope that this could be of help to anyone else who bumps into this issue and really wants to be on 5.* or other elasticsearch images which have auth enabled by default. :)

File: incubator/elasticsearch/templates/client-deployment.yaml (hunk 1 above)
Comment (rendhalver) [reply incubator/elasticsearch/templates/client-deployment.yaml:25]: I am not sure of the original reason for setting the gid here.
I will check and see if we actually need it.

**Comment by sescotti** [comment]
Just tested it with Elastic v6.1.1 on Kubernetes v1.9.7-gke.1 (GCP) and it worked like a charm!

**Comment by rendhalver** [comment]
@timstoop,@unguiculus, @simonswine and @icereval, do either of you have any other concerns with this?

**Review by unguiculus — APPROVED** [review]

**Comment by monotek** [comment]
@rendhalver there is a merge conflict again. could you please fix it?

**Comment by rendhalver** [comment]
Righto merge conflicts resolved.
Apologies for the late fix. I am currently enjoying the last day of a much needed week off.

**Comment by rendhalver** [comment]
Hooray!

**Comment by agolomoodysaada** [comment]
Testing latest Elastic 6.2.4 works well on AKS using k8s 1.9.6

**Comment by monotek** [comment]
Nice. Lets update the chart :D

**Comment by rendhalver** [comment]
We are using 6.2.1 for a couple of our clusters.

**Pull request merged**